docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,485 @@
+- r.BeforeCreate is the `rest.Creater` interface's new method for performing
+  logic after admission defaulting and before validation
+- r.Validate is the `rest.Creater` interface's new method for performing
+  validation
+- admit.Validate is a new admission phase that is executed after the
+  `rest.Creater`'s validation (plugins should not modify storage)
+- admit.EnforceConstraints is another new admission phase that is executed after
+  admit.Validate (plugins can modify storage)
+- admit.Finish (name TBD) is the final admission phase, running after the
+  `rest.Creater` has modified persistent storage
+
+
+### Update
+
+#### Current behavior
+
+1. decode request
+1. admit.Admit
+1. r.Update (the items below are part of the generic registry store / strategy-based storage, `pkg/registry/generic/registry/store.go`)
+  1. rest.BeforeUpdate
+    1. strategy.PrepareForUpdate
+    1. validation.ValidateObjectMetaUpdate
+    1. strategy.ValidateUpdate
+    1. strategy.Canonicalize
+  1. e.Storage.GuaranteedUpdate
+  1. e.AfterUpdate
+  1. e.Decorator
+
+#### Proposed behavior
+
+1. decode request
+1. ??? admit.ApplyDefaults ???
+1. r.BeforeUpdate
+  1. If strategy-based, call rest.BeforeUpdate
+    1. strategy.PrepareForUpdate
+  1. Otherwise, just do whatever is in the rest.Updater
+1. r.ValidateUpdate
+  1. If strategy-based, call rest.ValidateUpdate
+    1. validation.ValidateObjectMetaUpdate
+    1. strategy.ValidateUpdate
+    1. strategy.Canonicalize
+  1. Otherwise, just do whatever is in the rest.Updater
+1. admit.Validate
+1. admit.EnforceConstraints
+1. r.Update
+  1. For strategy-based handlers:
+    1. e.Storage.GuaranteedUpdate
+    1. e.AfterUpdate
+    1. e.Decorator
+  1. Otherwise, just do whatever is in the rest.Updater
+1. admit.Finish (name TBD)
+
+Notes:
+
+- Should we have admit.ApplyDefaults?
+- The `rest.Updater` interface's Update method has been split into 3:
+  - BeforeUpdate
+  - ValidateUpdate
+  - Update
+
+
+### Patch
+
+Patch is similar to Update, and similar changes are required to split the call
+to `Update` into `BeforeUpdate`, `ValidateUpdate`, and `Update`. Additionally,
+calls to `admit.Validate`, `admit.EnforceConstraints`, and `admit.Finish` will
+occur in appropriate places within the request lifecycle.
+
+### Delete
+
+#### Current behavior
+
+1. decode request (DeleteOptions)
+2. admit.Admit
+3. r.Delete (the items below are part of the generic registry store / strategy-based storage, `pkg/registry/generic/registry/store.go`)
+  4. e.Storage.Get
+  1. rest.BeforeDelete
+  1. e.Storage.Delete
+  1. e.AfterDelete
+  1. e.Decorator
+
+#### Proposed behavior
+
+1. decode request (DeleteOptions)
+2. r.Get
+1. r.BeforeDelete
+  2. If strategy-based, call rest.BeforeDelete
+  1. Otherwise, just do whatever is in the rest.Deleter
+1. admit.Validate
+1. admit.EnforceConstraints
+1. r.Delete
+  2. For strategy-based handlers:
+    1. e.Storage.Delete
+    1. e.AfterDelete
+    1. e.Decorator
+2. admit.Finish
```

Comment:
I think we need to guarantee it runs at least once. I need to do some more research about how we handle graceful termination and finalizers before I can answer your first question.